api/keyconjurer/serverless_functions.go

```diff
@@ -1,130 +1,232 @@
-	keyConjurerSettings, err := settings.NewSettings(logger)
-	if err != nil {
-		return CreateResponseError(err.Error()), err
+var (
+	errTimeoutBadSize = errors.New("ttl must be at least 1 hour and less than 8 hours")
+	errNoRoleProvided = errors.New("a role must be specified when using this identity provider")
+)
+
+// Validate validates that the event has appropriate parameters
+func (e GetTemporaryCredentialEvent) Validate() error {
+	if e.TimeoutInHours < 1 || e.TimeoutInHours > 8 {
+		return errTimeoutBadSize
 	}
 
-	auth, err := newAuthenticator(logger, keyConjurerSettings)
-	if err != nil {
-		return CreateResponseError(err.Error()), err
+	if e.RoleName == "" && e.AuthenticationProvider == AuthenticationProviderOkta {
+		return errNoRoleProvided
 	}
 
-	// make new keyconjurer instance
-	client, err := NewKeyConjurer(auth, logger, keyConjurerSettings)
+	return nil
+}
+
+type GetTemporaryCredentialsPayload struct {
+	// TODO: add CloudProvider property so the client can discriminate between different cloud providers
+	AccountID       string `json:"AccountId"` // Intentionally lower-cased to maintain backwards compatibilty
+	AccessKeyID     string `json:"AccessKeyId"`
+	SecretAccessKey string `json:"SecretAccessKey"`
+	SessionToken    string `json:"SessionToken"`
+	Expiration      string `json:"Expiration"`
+}
+
+// GetTemporaryCredentialEventHandler issues temporary credentials for the current user.
+//
+// This MUST be backwards compatible with the old version of KeyConjurer for a time.
+func (h *Handler) GetTemporaryCredentialEventHandler(ctx context.Context, event GetTemporaryCredentialEvent) (Response, error) {
+	if err := event.Validate(); err != nil {
+		return ErrorResponse(ErrBadRequest, err.Error())
+	}
+
+	creds := event.Credentials
+	provider, ok := h.authenticationProviders.Get(event.AuthenticationProvider)
+	if !ok {
+		return ErrorResponse(ErrCodeInvalidProvider, "invalid provider")
+	}
+
+	if err := h.crypt.Decrypt(ctx, &creds); err != nil {
+		return ErrorResponse(ErrCodeUnableToDecrypt, "unable to decrypt credentials")
+	}
+
+	response, err := provider.GenerateSAMLAssertion(ctx, creds, event.AppID)
 	if err != nil {
-		return CreateResponseError(err.Error()), err
+		return ErrorResponse(ErrCodeInternalServerError, "unable to generate SAML assertion")
 	}
 
-	user, err := client.providerClient.GetUserCredentials(event.Username, event.Password)
+	sts, err := h.aws.GetTemporaryCredentialsForUser(ctx, event.RoleName, response, int(event.TimeoutInHours))
 	if err != nil {
-		return CreateResponseError(err.Error()), errors.New("unable to get user information with current provider")
+		var errRoleNotFound aws.ErrRoleNotFound
+		if errors.As(err, &errRoleNotFound) {
+			return ErrorResponse(ErrBadRequest, errRoleNotFound.Error())
+		}
+
+		return ErrorResponse(ErrCodeInternalServerError, err.Error())
 	}
 
-	// Set the username field permanently for future logs
-	client.Logger = client.Logger.WithFields(logrus.Fields{
```

Comment:
As above, it was not obvious what client.Logger did from this code; the logging for credential generation should be done within the handler, and that was why when I wrote this I removed this. Although as we are creating temporary STS credentials, I wonder if there isn't a way that we can use CloudWatch for this data rather than KC? In the short term, we should reinstate the logging here